nixos/modules/system/service/portable/test.nix

```diff
@@ -0,0 +1,170 @@
+# Run:
+#   nix-instantiate --eval nixos/modules/system/service/portable/test.nix
+let
+  lib = import ../../../../../lib;
+
+  inherit (lib) mkOption types;
+
+  portable-lib = import ./lib.nix { inherit lib; };
+
+  dummyPkg =
+    name:
+    derivation {
+      system = "dummy";
+      name = name;
+      builder = "/bin/false";
+    };
+
+  exampleConfig = {
+    _file = "${__curPos.file}:${toString __curPos.line}";
+    services = {
+      service1 = {
+        process = {
+          executable = "/usr/bin/echo"; # *giggles*
+          args = [ "hello" ];
+        };
+        assertions = [
+          {
+            assertion = false;
+            message = "you can't enable this for that reason";
+          }
+        ];
+        warnings = [
+          "The `foo' service is deprecated and will go away soon!"
+        ];
+      };
+      service2 = {
+        process = {
+          # No meta.mainProgram, because it's supposedly an executable script _file_,
+          # not a directory with a bin directory containing the main program.
+          executable = dummyPkg "cowsay.sh";
+          args = [ "world" ];
+        };
+      };
+      service3 = {
+        process = {
+          executable = "/bin/false";
+          args = [ ];
+        };
+        services.exclacow = {
+          process = {
+            executable = dummyPkg "cowsay-ng" // {
+              meta.mainProgram = "cowsay";
+            };
+            args = [ "!" ];
+          };
+          assertions = [
+            {
+              assertion = false;
+              message = "you can't enable this for such reason";
+            }
+          ];
+          warnings = [
+            "The `bar' service is deprecated and will go away soon!"
+          ];
+        };
+      };
+    };
+  };
+
+  exampleEval = lib.evalModules {
+    modules = [
+      {
+        options.services = mkOption {
+          type = types.attrsOf (
+            types.submoduleWith {
+              class = "service";
```

Comment:
Noob question: what does `class` do?

Reply:
It's very basic "type checking" so that you don't load a normal NixOS module where a service is expected or vice versa.
The typing is "gradual" - if you don't specify it in a module or submodule, that counts as a match. In practice, this means modules have to specify `_class` to get this benefit. It's not so widely used yet. I think it's a 2 or 3 year old feature, but limited usefulness and not widely advertized. Flake-parts [`flake.modules`](https://flake.parts/options/flake-parts-modules.html) does set it, but that's more recent.
See also [`evalModules` `class`](https://nixos.org/manual/nixpkgs/stable/index.html#module-system-lib-evalModules-param-class) in the manual.

Reply:
Makes sense now. The manual helped also, specifically:

> If the class attribute is set and non-null, the module system will reject imports with a different _class declaration.